s3/src/main/scala/geotrellis/spark/io/s3/S3RDDReader.scala

```diff
@@ -45,37 +45,42 @@ trait S3RDDReader {
     val _getS3Client = getS3Client
     val kwWriterSchema = KryoWrapper(writerSchema) //Avro Schema is not Serializable
 
-    val rdd =
-      sc.parallelize(bins, bins.size)
-        .mapPartitions { partition: Iterator[Seq[(Long, Long)]] =>
-          val s3client = _getS3Client()
+    sc.parallelize(bins, bins.size)
+      .mapPartitions { partition: Iterator[Seq[(Long, Long)]] =>
+        val s3client = _getS3Client()
+
+        partition flatMap { seq =>
+          val range: Process[Task, Iterator[Long]] = Process.unfold(seq.toIterator) { iter =>
+            if (iter.hasNext) {
+              val (start, end) = iter.next()
+              Some((start to end).toIterator, iter)
+            } else None
+          }
 
-          val tileSeq: Iterator[Seq[(K, V)]] =
-            for {
-              rangeList <- partition // Unpack the one element of this partition, the rangeList.
-              range <- rangeList
-              index <- range._1 to range._2
-            } yield {
-              val path = keyPath(index)
-              val getS3Bytes = () => IOUtils.toByteArray(s3client.getObject(bucket, path).getObjectContent)
+          val read: Iterator[Long] => Process[Task, Vector[(K, V)]] = { iterator =>
+            Process.unfold(iterator) { iter =>
+              if (iter.hasNext) {
+                val index = iter.next()
+                val path = keyPath(index)
+                val getS3Bytes = () => IOUtils.toByteArray(s3client.getObject(bucket, path).getObjectContent)
 
-              try {
-                val bytes: Array[Byte] =
-                  getS3Bytes()
-                val recs = AvroEncoder.fromBinary(kwWriterSchema.value.getOrElse(_recordCodec.schema), bytes)(_recordCodec)
-                if(filterIndexOnly)
-                  recs
-                else
-                  recs.filter { row => includeKey(row._1) }
-              } catch {
-                case e: AmazonS3Exception if e.getStatusCode == 404 => Seq.empty
+                try {
+                  val bytes: Array[Byte] = getS3Bytes()
+                  val recs = AvroEncoder.fromBinary(kwWriterSchema.value.getOrElse(_recordCodec.schema), bytes)(_recordCodec)
+                  if (filterIndexOnly) Some(recs, iter)
+                  else Some(recs.filter { row => includeKey(row._1) }, iter)
+                } catch {
+                  case e: AmazonS3Exception if e.getStatusCode == 404 => Some(Vector.empty, iter)
+                }
+              } else {
+                None
               }
             }
+          }
 
-          tileSeq.flatten
+          nondeterminism.njoin(maxOpen = 8, maxQueued = 8) { range map read }.runFoldMap(identity).unsafePerformSync
```

Comment:
How do you feel about using typesafe config for this? This is probably a system wide property rather than per-action property.